crates/revm/src/db/alloydb.rs

```diff
@@ -0,0 +1,180 @@
+    pub fn new(provider: P, block_number: Option<BlockId>) -> Self {
+        Self {
+            provider,
+            block_number,
+            _marker: std::marker::PhantomData,
+        }
+    }
+
+    /// Internal utility function that allows us to block on a future regardless of the runtime flavor.
+    #[inline]
+    fn block_on<F>(f: F) -> F::Output
+    where
+        F: std::future::Future + Send,
+        F::Output: Send,
+    {
+        match Handle::try_current() {
+            Ok(handle) => match handle.runtime_flavor() {
+                // This is essentially equal to tokio::task::spawn_blocking because tokio doesn't
+                // allow the current_thread runtime to block_in_place.
+                // See <https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html> for more info.
+                tokio::runtime::RuntimeFlavor::CurrentThread => std::thread::scope(move |s| {
+                    s.spawn(move || {
+                        Builder::new_current_thread()
+                            .enable_all()
+                            .build()
+                            .unwrap()
+                            .block_on(f)
+                    })
+                    .join()
+                    .unwrap()
+                }),
+                _ => tokio::task::block_in_place(move || handle.block_on(f)),
+            },
+            Err(_) => Builder::new_current_thread()
+                .enable_all()
+                .build()
+                .unwrap()
+                .block_on(f),
+        }
+    }
+
+    /// Set the block number on which the queries will be based on.
+    pub fn set_block_number(&mut self, block_number: Option<BlockId>) {
+        self.block_number = block_number;
+    }
+}
+
+impl<T: Transport + Clone, N: Network, P: Provider<T, N>> DatabaseRef for AlloyDB<T, N, P> {
+    type Error = TransportError;
+
+    fn basic_ref(&self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        let f = async {
+            let nonce = self
+                .provider
+                .get_transaction_count(address, self.block_number);
+            let balance = self.provider.get_balance(address, self.block_number);
+            let code = self
+                .provider
+                .get_code_at(address, self.block_number.unwrap_or_default());
+            tokio::join!(nonce, balance, code)
+        };
+
+        let (nonce, balance, code) = Self::block_on(f);
+
+        let balance = balance?;
+        let code = Bytecode::new_raw(code?.0.into());
+        let code_hash = code.hash_slow();
+        let nonce = nonce?;
+
+        Ok(Some(AccountInfo::new(
+            balance,
+            nonce.to::<u64>(),
+            code_hash,
+            code,
+        )))
+    }
+
+    fn block_hash_ref(&self, number: U256) -> Result<B256, Self::Error> {
+        // Saturate usize
+        if number > U256::from(u64::MAX) {
```

Comment:
Wonder if need to determine block range. Since BLOCKHASH opcode semantics:
```
hash of the chosen block, or 0 if the block number is not in the valid range.
```

Reply:
@rakita

Reply:
This is done inside instruction